moto/core/models.py

```diff
@@ -124,31 +124,90 @@ def disable_patching(self):
                      responses.OPTIONS, responses.PATCH, responses.POST, responses.PUT]
 
 
-class ResponsesMockAWS(BaseMockAWS):
+class CallbackResponse(responses.CallbackResponse):
+    '''
+    Need to subclass so we can change a couple things
+    '''
+    def get_response(self, request):
+        '''
+        Need to override this so we can pass decode_content=False
+        '''
+        headers = self.get_headers()
+
+        result = self.callback(request)
+        if isinstance(result, Exception):
+            raise result
+
+        status, r_headers, body = result
+        body = responses._handle_body(body)
+        headers.update(r_headers)
+
+        return responses.HTTPResponse(
+            status=status,
+            reason=six.moves.http_client.responses.get(status),
+            body=body,
+            headers=headers,
+            preload_content=False,
+            # Need to not decode_content to mimic requests
+            decode_content=False,
+        )
+
+    def _url_matches(self, url, other, match_querystring=False):
+        '''
+        Need to override this so we can fix querystrings breaking regex matching
+        '''
+        if not match_querystring:
+            other = other.split('?', 1)[0]
+
+        if responses._is_string(url):
+            if responses._has_unicode(url):
+                url = responses._clean_unicode(url)
+                if not isinstance(other, six.text_type):
+                    other = other.encode('ascii').decode('utf8')
+            return self._url_matches_strict(url, other)
+        elif isinstance(url, responses.Pattern) and url.match(other):
+            return True
+        else:
+            return False
+
 
+botocore_mock = responses.RequestsMock(assert_all_requests_are_fired=False, target='botocore.vendored.requests.adapters.HTTPAdapter.send')
+
+
+class ResponsesMockAWS(BaseMockAWS):
     def reset(self):
+        botocore_mock.reset()
```

Comment:
What happens if we have multiple `ResponsesMockAWS` instances? The `RequestsMock` does not have an internal usage counter, so calling `reset` or `disable_patching` on any `ResponsesMockAWS` instance will reset or stop the `botocore_mock` for all.

Consider:

```
with mock_s3():
    with mock_ec2():
        ...
    with mock_ec2():
        ...
```
If I understand the workings of `moto` right, once leaving the scope of the first `mock_ec2` the `botocore_mock` is stopped.

Reply:
`reset` should only get called on the first decorator due to this block:

https://github.com/spulec/moto/blob/master/moto/core/models.py#L36-L37

I confirmed with your example code that it will only get called in the `mock_s3` one.

Reply:
You're right. My bad.